### ISSUE TITLE
Refactoring: reproducibility of the training pipeline

### ISSUE THREAD

**Issue by veegalinova** [CAREamics/careamics#460]
## Description
Currently, there's no way to ensure full reproducibility of the training pipeline. By "full reproducibility", I mean being able to exactly reproduce training loss values across multiple runs.

At the moment, the pipeline has two main sources of randomness: patch extraction and data transforms. All random operations are correctly using numpy's `default_rng()`. But, since the seed is None by default, each run will initialize the RNG differently, leading to variability between runs.

### Examples

- [patch extraction example](https://github.com/CAREamics/careamics/blob/488255829a3e95acde4252752b2f793ef950cf71/src/careamics/dataset/patching/sequential_patching.py#L136)
- [xy_flip transform example](https://github.com/CAREamics/careamics/blob/488255829a3e95acde4252752b2f793ef950cf71/src/careamics/transforms/xy_flip.py#L46)

``` python
rng = np.random.default_rng()  # seed not set, nor passed from the init
...
rng.shuffle(patches, axis=0) # will be randomized between runs
```

## Potential solutions
1) add a `random seed` parameter to the data configuration (and maybe consider setting a value by default?)
2) propagate this parameter through the pipeline to the patching and to the transforms

## Additional context 
For clarity, to achieve complete reproducibility, two more settings need to be handled outside the repo's code:
1) setting the seeds for each of the workers, can be done with lightning:
```python
from pytorch_lightning import seed_everything
seed_everything(42)
```
2) enabling torch determinism (may slightly impact performance):
```
torch.use_deterministic_algorithms(True)
torch.backends.cudnn.benchmark = False
```

We also need to insure that it will not affect the performance of current algorithms.

**Comment by jdeschamps** [comment]
So the solutions for fully reproducibility would be:
1. `random_seed` parameter in the `DataConfig`
2. a convenience function to set reproducibility mode, which can for instance be called by `CAREamist`

**Comment by veegalinova** [comment]
Two more things I realized since then: 
1. We need to ensure that each worker has the same random seed, which should be handled by lightning's `seed_everything`, but need to double-check.
2. The datasets are created each epoch, so if the random seed is set, the samples will have the same augmentations each epoch. Not sure if it's critical for performance, need to test.

**Comment by melisande-c** [comment]
> 2. The datasets are created each epoch, so if the random seed is set, the samples will have the same augmentations each epoch. Not sure if it's critical for performance, need to test.

I'm pretty sure that the `setup` hook in the datamodule, where the datasets are initialised, is only called at the start of each stage, so the training dataset will only be initialised once, at the start of training. And, in the case of multiple workers, a new copy of the dataset will be given to each worker

**Comment by veegalinova** [comment]
You are right! It was somehow not clear to me from the docs. I thought the training internally should alternate between fit and validate stages and so re-initialize, which is not the case, it's just fit stage all the way through. 
It will only re-initialize if a stage is called manually like `trainer.fit()` or `trainer.validate()`.

**Comment by jdeschamps** [comment]
@veegalinova has this mostly been implemented now with the automated seeding of the configuration?
